toolchain/lexer/tokenized_buffer_benchmark.cpp

```diff
@@ -0,0 +1,136 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <benchmark/benchmark.h>
+
+#include <forward_list>
+
+#include "absl/random/random.h"
+#include "common/check.h"
+#include "llvm/ADT/Sequence.h"
+#include "llvm/ADT/StringExtras.h"
+#include "toolchain/diagnostics/null_diagnostics.h"
+#include "toolchain/lexer/token_kind.h"
+#include "toolchain/lexer/tokenized_buffer.h"
+
+namespace Carbon::Testing {
+namespace {
+
+class LexerBenchHelper {
+ public:
+  explicit LexerBenchHelper(llvm::StringRef text)
+      : source_(MakeSourceBuffer(text)) {}
+
+  auto Lex() -> TokenizedBuffer {
+    DiagnosticConsumer& consumer = NullDiagnosticConsumer();
+    return TokenizedBuffer::Lex(source_, consumer);
+  }
+
+  auto DiagnoseErrors() -> llvm::StringRef {
+    auto buffer = TokenizedBuffer::Lex(source_, ConsoleDiagnosticConsumer());
+    CARBON_CHECK(buffer.has_errors())
+        << "Asked to diagnose errors but none found!";
+    // Return an empty string so this can be streamed into a check failure.
+    return "";
+  }
+
+ private:
+  llvm::vfs::InMemoryFileSystem fs_;
+  std::string filename_ = "test.carbon";
+  SourceBuffer source_;
+
+  auto MakeSourceBuffer(llvm::StringRef text) -> SourceBuffer {
+    CARBON_CHECK(fs_.addFile(filename_, /*ModificationTime=*/0,
+                             llvm::MemoryBuffer::getMemBuffer(text)));
+    return std::move(*SourceBuffer::CreateFromFile(fs_, filename_));
+  }
+};
+
+constexpr int NumTokens = 100000;
+
+void BM_ValidKeywords(benchmark::State& state) {
+  absl::BitGen gen;
+  std::string source;
+  llvm::raw_string_ostream os(source);
+  llvm::ListSeparator sep(" ");
+  for (int i = 0; i < NumTokens; ++i) {
+    int token = absl::Uniform<int>(gen, 0, TokenKind::KeywordTokens.size());
+    os << sep << TokenKind::KeywordTokens[token].fixed_spelling();
+  }
+
+  LexerBenchHelper helper(source);
+  for (auto _ : state) {
+    TokenizedBuffer buffer = helper.Lex();
+    benchmark::DoNotOptimize(buffer.has_errors());
+  }
+
+  state.counters["TokenRate"] = benchmark::Counter(
+      NumTokens, benchmark::Counter::kIsIterationInvariantRate);
+}
+BENCHMARK(BM_ValidKeywords);
+
+// clang-format off
+constexpr char IdentifierChars[] = {
+    // Digits:
+    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
+    // Upper case:
+    'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O',
+    'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z',
+    // Lower case:
+    'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'M', 'n', 'o',
```

Comment:
```suggestion
    'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o',
```

Might be worth generating by code to reduce visual validation?

Reply:
Applied the fix.

Note that it is *very* hard to make a `constexpr` variable that is generated by code. I tried, hard, and after a bunch of effort gave up.

I guess I'll change this to not be `constexpr`, and instead be a lazily computed global with a bunch of heap allocations behind it. Not seeing better options here, which is why I initially went with just a `constexpr` array.

Done, but made all the code a bit longer.